src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java

```diff
@@ -15,55 +14,67 @@
 import org.apache.parquet.schema.MessageType;
 import org.apache.parquet.schema.Type;
 
-/** Helper class to find all leaf sub-columns in an immutable schema given a dot path. */
+/** Helper class to find all leaf sub-columns in an immutable schema given a fieldId. */
 public class SubColumnFinder {
-  static class SubtreeInterval {
+  static class SubtreeInfo {
     final int startTag;
     final int endTag;
+    final String dotPath;
 
-    SubtreeInterval(int startTag, int endTag) {
+    SubtreeInfo(int startTag, int endTag, String dotPath) {
       this.startTag = startTag;
       this.endTag = endTag;
+      this.dotPath = dotPath;
     }
   }
 
   private final List<String> list;
-  private final Map<String, SubtreeInterval> accessMap;
+  private final Map<String, SubtreeInfo> accessMap;
 
   public SubColumnFinder(MessageType schema) {
     accessMap = new HashMap<>();
     list = new ArrayList<>();
     build(schema, null);
   }
 
-  public List<String> getSubColumns(String dotPath) {
-    if (!accessMap.containsKey(dotPath)) {
-      throw new IllegalArgumentException(String.format("Column %s not found in schema", dotPath));
+  public List<String> getSubColumns(String id) {
+    if (!accessMap.containsKey(id)) {
+      throw new IllegalArgumentException(String.format("Field %s not found in schema", id));
     }
-    SubtreeInterval interval = accessMap.get(dotPath);
+    SubtreeInfo interval = accessMap.get(id);
     return Collections.unmodifiableList(list.subList(interval.startTag, interval.endTag));
   }
 
-  private void build(Type node, String dotPath) {
-    if (dotPath == null) {
+  public String getDotPath(String id) {
+    if (!accessMap.containsKey(id)) {
+      throw new IllegalArgumentException(String.format("Field %s not found in schema", id));
+    }
+    return accessMap.get(id).dotPath;
+  }
+
+  private void build(Type node, List<String> path) {
+    if (path == null) {
       /* Ignore root node type name (bdec or schema) */
-      dotPath = "";
-    } else if (dotPath.isEmpty()) {
-      dotPath = node.getName();
+      path = new ArrayList<>();
     } else {
-      dotPath = concatDotPath(dotPath, node.getName());
+      path.add(node.getName());
     }
 
     int startTag = list.size();
     if (!node.isPrimitive()) {
       for (Type child : node.asGroupType().getFields()) {
-        build(child, dotPath);
+        build(child, path);
       }
     } else {
-      list.add(dotPath);
+      list.add(node.getId().toString());
+    }
+    if (!path.isEmpty() && node.getId() != null) {
+      accessMap.put(
```

Comment:
we should change accessMap's type to be Map<ID, SubtreeInfo> instead of String.
I see every place doing ID.toString(), might as well use strong typing to enforce the ID is what comes in and not just any String.

Reply:
Switch to `Type.ID`